### PR TITLE
cf-coding

### PR DESCRIPTION
**Pull request by kmuehlbauer**
- [x] xref #7652
- [x] Tests added
- [ ] User visible changes (including notable bug fixes) are documented in `whats-new.rst`
- [ ] New functions/methods are listed in `api.rst`
- includes
    - implement coders
    - preserve boolean-dtype within encoding, adapt test
    - determine cf packed data type
    - ~transform numpy object-dtype strings (vlen) to numpy unicode strings~

This should also fix:

- Closes https://github.com/pydata/xarray/issues/7691
- Closes https://github.com/pydata/xarray/issues/2304
and possible other issues which have the float32/float64 issue

### CODE REVIEW

**Comment by kmuehlbauer** [comment]
I'll add to `whats-new.rst` if an updated version is merged.

**Comment by kmuehlbauer** [comment]
The one failed run might be spurious. Maybe a re-run of this would work.

**Comment by basnijholt** [comment]
I can confirm that this fixes
- https://github.com/pydata/xarray/issues/7652#issuecomment-1476956975 (`bool` -> `int8`)

But not `int64` -> `int32`, and `<U1` -> `O`.

**Comment by kmuehlbauer** [comment]
@basnijholt Thanks for testing. I can't reproduce this. Here everything works as expected. But I've a slightly different environment (full netcdf4-stack, latest versions of everything). 

The tests which check for dtype equality (vlen-case) do not raise in this PR for any backend and array container, so I'm assuming this should work as expected (at least for `bool`->`int8`, and `<U1` -> `O`.

As stated over at #7652 I can't reproduce the int64->int32 conversion in any environments I tested so far. I'll have another look at your environment.

File: xarray/tests/test_backends.py [hunk 1]
```diff
@@ -632,6 +632,9 @@ def test_roundtrip_boolean_dtype(self) -> None:
         with self.roundtrip(original) as actual:
             assert_identical(original, actual)
             assert actual["x"].dtype == "bool"
+            with self.roundtrip(actual) as actual2:
```
Comment (kmuehlbauer) [inline xarray/tests/test_backends.py:635]: This checks for preserving dtype during second roundtrip.

File: xarray/conventions.py [hunk 2]
```diff
@@ -389,6 +389,15 @@ def decode_cf_variable(
     if decode_times:
         var = times.CFDatetimeCoder(use_cftime=use_cftime).decode(var, name=name)
 
+    # transform numpy object-dtype strings to numpy unicode strings
```
Comment (kmuehlbauer) [inline xarray/conventions.py:392]: I'm not absolutely sure, if this is really what we want. This cast's the variable to the _expected_ dtype, but I can't immediately see the implications for users downstream.

File: xarray/tests/test_backends.py [hunk 3]
```diff
@@ -1357,6 +1360,7 @@ def test_encoding_kwarg_vlen_string(self) -> None:
             with self.roundtrip(original, save_kwargs=kwargs) as actual:
                 assert actual["x"].encoding["dtype"] is str
                 assert_identical(actual, expected)
+                assert actual["x"].dtype == "<U3"
```
Comment (kmuehlbauer) [inline xarray/tests/test_backends.py:1363]: Checks for preserved dtype.

**Comment by kmuehlbauer** [comment]
> But **not** `int64` -> `int32`, and `<U1` -> `O`

@basnijholt I've explained the issues whereabouts over at your issue https://github.com/pydata/xarray/issues/7652#issuecomment-1479080036.

**Comment by kmuehlbauer** [comment]
XRef: https://github.com/pydata/xarray/issues/2040#issuecomment-379418732

Citing @shoyer from above comment:

> The main reason why we don't do any special handling for object arrays currently in xarray is that our conventions coding/decoding system has no way of marking variable length string arrays. We should probably handle this by making a custom dtype like h5py that marks variables length strings using dtype metadata: http://docs.h5py.org/en/latest/special.html#variable-length-strings

Another interesting issue by @shoyer:
https://github.com/pydata/xarray/issues/2059

I'm really uncertain if using `.astype` here is the right way to go. Any comments appreciated.

**Comment by kmuehlbauer** [comment]
And yet another related issue: https://github.com/pydata/xarray/issues/1647

Currently both netcdf/h5netcdf are able to set a `_FillValue` for VLEN strings, but for the numpy-side "NaN" can only be handled with dtype=object. Maybe it's time to consolidate string handling in xarray. But that should be taken care of in a separate issue / feature branch.

**Comment by kmuehlbauer** [comment]
> make

That's 15 UTC, or 17 CEST (my local time). Should work for me. I'll try to collect all available information on that topic and the current status.

**Comment by basnijholt** [comment]
Just leaving a note here. I would expect that the datatype that was saved, is the datatype that is loaded. So preferably if I save a string array of e.g., type `<U5`, I expect it would still be `<U5` when loaded, not suddenly HDF5 `VLEN` types.

Thanks again @kmuehlbauer for digging into this problem and all your work! 😄

**Comment by kmuehlbauer** [comment]
@dcherian @basnijholt 

After the dev-meeting I've taken a step back and first implemented the coders as mentioned in @shoyer's ToDo.  

I've fixed the one bool->int issue and it now derives the dtype for ScaleOffset coding from scale_factor add_offset.

I've improved some test with regard to the scale/offset issue.

I'll concentrate on the string fillvalue issues in a follow up PR.

File: xarray/tests/test_backends.py [hunk 4]
```diff
@@ -138,96 +138,96 @@ def open_example_mfdataset(names, *args, **kwargs) -> Dataset:
     )
 
 
-def create_masked_and_scaled_data() -> Dataset:
-    x = np.array([np.nan, np.nan, 10, 10.1, 10.2], dtype=np.float32)
+def create_masked_and_scaled_data(dtype=np.float32) -> Dataset:
```
Comment (Illviljan) [inline xarray/tests/test_backends.py:141]:
```suggestion
def create_masked_and_scaled_data(dtype: np.typing.DTypeLike = np.float32) -> Dataset:
```

Can do this in other places too.

File: xarray/coding/variables.py [hunk 5]
```diff
@@ -232,25 +232,56 @@ def _scale_offset_decoding(data, scale_factor, add_offset, dtype: np.typing.DTyp
...
+        dims, data, attrs, encoding = unpack_for_encoding(variable)
+        # make NaN the fill value for float types
+        if (
+            "_FillValue" not in attrs
+            and "_FillValue" not in encoding
+            and np.issubdtype(variable.dtype, np.floating)
+        ):
+            attrs["_FillValue"] = variable.dtype.type(np.nan)
+
+        return Variable(dims, data, attrs, encoding, fastpath=True)
```
Comment (Illviljan) [inline xarray/coding/variables.py:279]:
```suggestion
        if (
            "_FillValue" not in attrs
            and "_FillValue" not in encoding
            and np.issubdtype(variable.dtype, np.floating)
        ):
            attrs["_FillValue"] = variable.dtype.type(np.nan)
            return Variable(dims, data, attrs, encoding, fastpath=True)
        else:
            return variable
```
No need to create a new Variable if nothing was changed.

File: xarray/coding/variables.py [hunk 6]
```diff
@@ -349,3 +380,144 @@ def decode(self, variable: Variable, name: T_Name = None) -> Variable:
...
+
+    def __init__(self, array):
+        self.array = indexing.as_indexable(array)
+
+    @property
+    def dtype(self):
+        return np.dtype(self.array.dtype.kind + str(self.array.dtype.itemsize))
+
+    def __getitem__(self, key):
+        return np.asarray(self.array[key], dtype=self.dtype)
```
Comment (Illviljan) [inline xarray/coding/variables.py:475]: This converts to numpy. Are we sure this is what we want?

For more discussions, see #6874

File: xarray/coding/variables.py [hunk 7]
```diff
@@ -349,3 +380,144 @@ def decode(self, variable: Variable, name: T_Name = None) -> Variable:
...
+
+    def __init__(self, array):
+        self.array = indexing.as_indexable(array)
+
+    @property
+    def dtype(self):
+        return np.dtype("bool")
+
+    def __getitem__(self, key):
+        return np.asarray(self.array[key], dtype=self.dtype)
```
Comment (Illviljan) [inline xarray/coding/variables.py:414]: This converts to numpy. Are we sure this is what we want?

For more discussions, see #6874

File: xarray/coding/variables.py (hunk 6 above)
Comment (kmuehlbauer) [reply xarray/coding/variables.py:475]: Thanks @Illviljan, I'll check how to make this lazy too. Any suggestions are greatly appreciated.

Comment (dcherian) [reply xarray/coding/variables.py:475]: Hmmm... This diff is misleading I think. This was moved from `conventions.py`? I'll merge #6874 and then I think resolving the merge conflicts should make it OK.

Comment (dcherian) [reply xarray/coding/variables.py:475]: Oh  wait that PR didn't change `conventions.py` and the comment mentions netCDF3 for which #6874 isn't relevant.

I would just ignore this till it becomes a problem. It's orthogonal to this PR anyway.

Comment (kmuehlbauer) [reply xarray/coding/variables.py:475]: I'll have a look what is going on here. I've very much tampered with branches and force pushes. Good chance I made something wrong.

File: xarray/conventions.py (hunk 2 above)
Comment (kmuehlbauer) [reply xarray/conventions.py]: removed from PR

**Comment by kmuehlbauer** [comment]
@dcherian @Illviljan Thanks for the first round of review. I've rebased everything on latest main. Now the code moving from `conventions.py` to `coding.variable.py` is correct. I've also removed the functions which have been converted to `VariableCoders` and adapted the tests.

To sum up this PR, it does:

- convert functions to `VariableCoders` along @shoyers TODO:
https://github.com/pydata/xarray/blob/1c81162755457b3f4dc1f551f0321c75ec9daf6c/xarray/conventions.py#L298-L302
https://github.com/pydata/xarray/blob/1c81162755457b3f4dc1f551f0321c75ec9daf6c/xarray/conventions.py#L393-L405
- preserve boolean dtype within `encoding`: 
https://github.com/pydata/xarray/issues/7652#issuecomment-1476956975
- deterrmine cf packed dtype from `scale_factor`/`add_offset`
#7691, #2304

**Comment by kmuehlbauer** [comment]
@Illviljan I'm not able to figure out the typing if I want to use Data-types as functions to convert python numbers to array scalars. If you have any suggestion how to fix this, please let me know.

File: xarray/coding/variables.py [hunk 8]
```diff
@@ -232,20 +297,32 @@ def _scale_offset_decoding(data, scale_factor, add_offset, dtype: np.typing.DTyp
     return data
 
 
-def _choose_float_dtype(dtype: np.dtype, has_offset: bool) -> type[np.floating[Any]]:
-    """Return a float dtype that can losslessly represent `dtype` values."""
-    # Keep float32 as-is.  Upcast half-precision to single-precision,
+def _choose_float_dtype(
```
Comment (dcherian) [inline xarray/coding/variables.py:300]: @mankoff do you have time to take a look here please?

File: xarray/tests/test_backends.py (hunk 1 above)
Comment (dcherian) [reply xarray/tests/test_backends.py:647]: Lets add this as a comment

File: xarray/tests/test_backends.py [hunk 9]
```diff
@@ -873,12 +891,19 @@ def test_roundtrip_string_with_fill_value_nchar(self) -> None:
...
+        ):
+            pytest.skip(
+                "zarr attributes (eg. `scale_factor` are unconditionally promoted to `float64`"
+            )
+        decoded = decoded_fn(dtype)
+        encoded = encoded_fn(dtype)
 
         with self.roundtrip(decoded) as actual:
             for k in decoded.variables:
+                print(k, decoded.variables[k].dtype)
```
Comment (dcherian) [inline xarray/tests/test_backends.py:906]:
```suggestion
```

File: xarray/coding/variables.py [hunk 10]
```diff
@@ -232,20 +297,32 @@ def _scale_offset_decoding(data, scale_factor, add_offset, dtype: np.typing.DTyp
...
+    # check scale/offset first to derive dtype with
+    if "scale_factor" in encoding or "add_offset" in encoding:
+        scale_factor = encoding.get("scale_factor", False)
+        add_offset = encoding.get("add_offset", False)
+        # minimal floating point size -> 4 byte
+        maxsize = 4
+        if scale_factor and np.issubdtype(type(scale_factor), np.floating):
+            maxsize = max(maxsize, np.dtype(type(scale_factor)).itemsize)
+        if add_offset and np.issubdtype(type(add_offset), np.floating):
+            maxsize = max(maxsize, np.dtype(type(add_offset)).itemsize)
```
Comment (dcherian) [inline xarray/coding/variables.py:312]: Some comments and links to Github issues / comments would be a good addition.

File: xarray/coding/variables.py (hunk 8 above)
Comment (mankoff) [reply xarray/coding/variables.py:300]: Hi @dcherian . I'm not sure what to look for. I will link to my open-but-stale PR that tried to start addressing this/similar issues: https://github.com/pydata/xarray/pull/6812

Perhaps also relevant are my last few comments on #2304 (see https://github.com/pydata/xarray/issues/2304#issuecomment-1201461626 ). The problem for me is that (1) the CF standards are ambiguously defined and (2) xarray needs to address the many use-cases where the CF standards are not followed (usually this means different data types).

Comment (dcherian) [reply xarray/coding/variables.py:300]: Thanks! Does this PR fix your original issue?

Comment (kmuehlbauer) [reply xarray/coding/variables.py:300]: Thanks @mankoff, I'll have a look at your extensive notes over there and try to come up with aomething.

File: xarray/tests/test_backends.py [hunk 11]
```diff
@@ -138,96 +138,110 @@ def open_example_mfdataset(names, *args, **kwargs) -> Dataset:
     )
 
 
-def create_masked_and_scaled_data() -> Dataset:
-    x = np.array([np.nan, np.nan, 10, 10.1, 10.2], dtype=np.float32)
+def create_masked_and_scaled_data(dtype: np.typing.DTypeLike = np.float32) -> Dataset:
```
Comment (Illviljan) [inline xarray/tests/test_backends.py:141]:
```suggestion
def create_masked_and_scaled_data(dtype: type[np.number] = np.float32) -> Dataset:
```

aha, this is not the standard dtype variable. 

This example passes mypy, so should hopefully work for these functions as well:
```python 
import numpy as np


def test(dtype: type[np.number] = np.float32) -> None:
    print(np.float32(32))

    if isinstance(dtype, type(np.number)):
        print(np.float32(32))


test(np.float32)

```

**Comment by kmuehlbauer** [comment]
The latest changes brake #1840 again. We have two contradicting forces here, which need to be aligned.

**Comment by kmuehlbauer** [comment]
If at first you don't succeed... It looks like we have something working here.

Some more typing and maybe some more tests covering the cases with scale_factor/add_offset/_FillValue non-conforming CF and we should be good to go. Or do I miss something?

File: xarray/tests/test_backends.py [hunk 12]
```diff
@@ -873,9 +893,17 @@ def test_roundtrip_string_with_fill_value_nchar(self) -> None:
             (create_masked_and_scaled_data, create_encoded_masked_and_scaled_data),
         ],
     )
-    def test_roundtrip_mask_and_scale(self, decoded_fn, encoded_fn) -> None:
-        decoded = decoded_fn()
-        encoded = encoded_fn()
+    def test_roundtrip_mask_and_scale(
+        self, decoded_fn: Callable, encoded_fn: Callable, dtype: type[np.number]
```
Comment (Illviljan) [inline xarray/tests/test_backends.py:897]:
```suggestion
        self, decoded_fn: Callable[[type[np.number]], Dataset], encoded_fn: Callable[[type[np.number]], Dataset], dtype: type[np.number]
```

File: xarray/tests/test_backends.py [hunk 13]
```diff
@@ -1530,31 +1558,76 @@ def test_encoding_chunksizes_unlimited(self) -> None:
...
             # now check xarray
             with open_dataset(tmp_file) as ds:
-                expected = create_masked_and_scaled_data()
+                expected = create_masked_and_scaled_data(dtype)
                 assert_identical(expected, ds)
 
+    @pytest.mark.parametrize("dtype", [np.float32, np.float64])
+    @pytest.mark.parametrize("offset_cf_conforming", [True, False])
+    def test_mask_and_scale_non_cf_conforming(
+        self, dtype, offset_cf_conforming
```
Comment (Illviljan) [inline xarray/tests/test_backends.py:1592]:
```suggestion
        self, dtype: type[np.number], offset_cf_conforming: bool
```

File: xarray/coding/variables.py [hunk 14]
```diff
@@ -349,3 +436,100 @@ def decode(self, variable: Variable, name: T_Name = None) -> Variable:
...
+
+
+class NonStringCoder(VariableCoder):
+    """Encode NonString variables if dtypes differ."""
+
+    def encode(self, variable: Variable, name: T_Name = None) -> Variable:
+        if "dtype" in variable.encoding and variable.encoding["dtype"] not in (
+            "S1",
+            str,
+        ):
```
Comment (Illviljan) [inline xarray/coding/variables.py:515]:
```suggestion
        if "dtype" in variable.encoding and variable.encoding["dtype"] not in ("S1", str):
```
Strange decision by black here... #7600

**Comment by kmuehlbauer** [comment]
This is now ready for another round of reviews, @dcherian, @Illviljan and @mankoff.

As @mankoff already pointed out, xarray is very generous to try to encode/decode non CF conforming data. This makes things a bit complicated as some issues only surface in rare corner cases.

I've tried to be as explicit in `_choose_float_dtype`, also added comments/tests where needed.

I'm finding the typing a bit hard. It seems that mypy can't derive the correct types from return types in certain cases.

File: xarray/coding/variables.py [hunk 15]
```diff
@@ -232,20 +297,46 @@ def _scale_offset_decoding(data, scale_factor, add_offset, dtype: np.typing.DTyp
...
+            offset_type = type(add_offset)
+        # CF conforming, both scale_factor and add-offset are given and
+        # of same floating point type
+        if (
+            add_offset
+            and scale_factor
+            and offset_type == scale_type
+            and np.issubdtype(scale_type, np.floating)
+        ):
+            return np.dtype(scale_type)
```
Comment (Illviljan) [inline xarray/coding/variables.py:322]:
```suggestion
            return np.dtype(scale_type).type
```
Probably want the same return type as the others.

```python
np.dtype(np.float32)
Out[6]: dtype('float32')

np.dtype(np.float32).type
Out[7]: numpy.float32

# Can't call dtypes:
np.dtype(np.float32)(1)
TypeError: 'numpy.dtype[float32]' object is not callable

np.dtype(np.float32).type(1)
Out[13]: 1.0
```

Comment (kmuehlbauer) [reply xarray/coding/variables.py]: Oh, my bad... Thanks @Illviljan, your help is much appreciated. BTW, the new batch committing in the GitHub UI is very comfortable.

File: xarray/coding/variables.py (hunk 14 above)
Comment (kmuehlbauer) [reply xarray/coding/variables.py:515]: Seems like pre-commit put's this back.

File: xarray/coding/variables.py [hunk 16]
```diff
@@ -78,6 +78,71 @@ def __repr__(self) -> str:
...
+
+    def __init__(self, array):
+        self.array = indexing.as_indexable(array)
+
+    @property
+    def dtype(self):
+        return np.dtype(self.array.dtype.kind + str(self.array.dtype.itemsize))
+
+    def __getitem__(self, key):
+        return np.asarray(self.array[key], dtype=self.dtype)
```
Comment (Illviljan) [inline xarray/coding/variables.py:111]:
```suggestion
    def __init__(self, array) -> None:
        self.array = indexing.as_indexable(array)

    @property
    def dtype(self) -> np.dtype:
        return np.dtype(self.array.dtype.kind + str(self.array.dtype.itemsize))

    def __getitem__(self, key) -> np.ndarray:
        return np.asarray(self.array[key], dtype=self.dtype)
```
Would be nice to type `array` and `key` but they're not as easy though.

File: xarray/coding/variables.py [hunk 17]
```diff
@@ -78,6 +78,71 @@ def __repr__(self) -> str:
...
+
+    def __init__(self, array):
+        self.array = indexing.as_indexable(array)
+
+    @property
+    def dtype(self):
+        return np.dtype("bool")
+
+    def __getitem__(self, key):
+        return np.asarray(self.array[key], dtype=self.dtype)
```
Comment (Illviljan) [inline xarray/coding/variables.py:143]:
```suggestion
    def __init__(self, array) -> None:
        self.array = indexing.as_indexable(array)

    @property
    def dtype(self) -> np.dtype:
        return np.dtype("bool")

    def __getitem__(self, key) -> np.ndarray:
        return np.asarray(self.array[key], dtype=self.dtype)
```

**Review by Illviljan — APPROVED** [review]
Looking good to me. :)

File: xarray/coding/variables.py (hunk 17 above)
Comment (kmuehlbauer) [reply xarray/coding/variables.py]: Thanks @Illviljan :heart:

**Comment by kmuehlbauer** [comment]
While trying to create a test which specifically tests `_choose_float_dtype` I've found some issues with checking for availability of scale_factor/add_offset. Now testing for `None`.

**Comment by kmuehlbauer** [comment]
@dcherian Just a heads-up: I find this PR getting more and more involved at different parts of the machinery and hard to follow for reviewers. I'll split this up and start with the more or less undisputed changes.

**Comment by kmuehlbauer** [comment]
As explained I've created two PR (#7719 and #7720) for the "easy" changes from this PR. Would be great, if those could go in fast. Thanks!

**Comment by kmuehlbauer** [comment]
I've converted to draft for now, as I'm still evaluating solutions for the CF encoding/decoding.

**Comment by kmuehlbauer** [comment]
I'll close this one. Most things have been addressed in other PR's.
